### PR TITLE
DDF-3024 CSV Query Response Transformer

### PR DESCRIPTION
**Pull request by dcruver**
#### What does this PR do?
This change creates an implementation of QueryResponseTransformer that translates the results from metacards to CSV format. Front-end support for this
will be added in a separate story.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@gordocanchola (hero)
@brianfelix 

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[UI](https://github.com/orgs/codice/teams/ui) (@andrewfielder)
[Data](https://github.com/orgs/codice/teams/data) (@brianfelix)

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@andrewkfiedler
@jlcsmith

#### How should this be tested? (List steps with links to updated documentation)
1) ingest some number of products.
2) perform an OpenSearch query and specify 'csv' as the output format 
     Example: curl --basic --user admin:admin --insecure "https://localhost:8993/services/catalog/query?q=*&format=csv"

#### What are the relevant tickets?
[DDF-3024](https://codice.atlassian.net/browse/DDF-3024)
#### Checklist:
- [ ] Documentation Updated
- [ ] Change Log Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

**Comment by dcruver** [comment]
build now

**Comment by dcruver** [comment]
build now

File: catalog/spatial/csv/spatial-csv-transformer/src/main/java/org/codice/ddf/spatial/ogc/csv/catalog/transformer/MetacardIterator.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        AttributeDescriptor attributeDescriptor = this.attributeDescriptorList.get(index);
+        Attribute attribute = metacard.getAttribute(attributeDescriptor.getName());
+        index++;
+
+        if (attribute != null) {
+            return attribute.getValue();
+        }
+
+        return "";
```
Comment (rzwiefel) [inline catalog/spatial/csv/spatial-csv-transformer/src/main/java/org/codice/ddf/spatial/ogc/csv/catalog/transformer/MetacardIterator.java]: I believe we have both the concept of an attribute value being `null` and `""` -- as two separate and distinct values. I don't think this will override some implicit behavior we have without returning null when the value is really null.

Comment (dcruver) [inline catalog/spatial/csv/spatial-csv-transformer/src/main/java/org/codice/ddf/spatial/ogc/csv/catalog/transformer/MetacardIterator.java]: 'attributeDescriptorList' is going to contain the union of all the AttributeDescriptors from a returned query because the output CSV has to contain a column for each one. `metacard.getAttribute("nitf.imageMode") is going to return an attribute for Nitf metacards, but not for FMV ones (for example). So we will frequently be getting null attributes returned and calling `.getValue()` on null is going to cause an NPE. The CSVPrinter seems to handle null and "" the same way, so I'm not sure that `return null` is better than `return ""` in the case where the attribute object is null.

In short, if the attribute's value is null then null will be returned.  If the attribute doesn't exist in the metacard, you'll get back "". And for the purpose of generating a CSV file, they're equivalent.

**Comment by rzwiefel** [comment]
build now

File: catalog/transformer/catalog-transformer-csv/src/test/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformerTest.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+        for (int i = 0; i < METACARD_COUNT; i++) {
+            Result result = mock(Result.class);
+            Metacard metacard = buildMetacard(i);
+            when(result.getMetacard()).thenReturn(metacard);
+            RESULT_LIST.add(result);
+        }
+    }
+
+    private Metacard buildMetacard(final int id) {
```
Comment (rzwiefel) [inline catalog/transformer/catalog-transformer-csv/src/test/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformerTest.java]: Codacity is complaining for you to remove this since it is never used, Though not sure why it didn't comment on the PR itself.

Comment (dcruver) [inline catalog/transformer/catalog-transformer-csv/src/test/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformerTest.java]: Removed.

File: catalog/core/catalog-core-solrcommands/pom.xml [hunk 3]
```diff
@@ -156,7 +156,7 @@
                 <artifactId>maven-bundle-plugin</artifactId>
                 <configuration>
                     <instructions>
-                        <Karaf-Commands>*</Karaf-Commands>
+                        <Karaf-Commands>org.codice.ddf.commands.solr</Karaf-Commands>
```
Comment (jlcsmith) [inline catalog/core/catalog-core-solrcommands/pom.xml]: Is this change related to this PR?

File: catalog/catalog-app/pom.xml [hunk 4]
```diff
@@ -594,5 +594,15 @@
...
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-csv</artifactId>
+            <version>${commons-csv.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>ddf.catalog.transformer</groupId>
+            <artifactId>csv-queryresponse-transformer</artifactId>
+            <version>2.11.0.pte.0-SNAPSHOT</version>
```
Comment (jlcsmith) [inline catalog/catalog-app/pom.xml]: please correct version

File: catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/MetacardIterator.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+/*
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/MetacardIterator.java]: remove duplicate header

File: catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformer.java [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+/*
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformer.java]: remove duplicate header

File: catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/CsvAttributeDescriptorComparator.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+/**
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/CsvAttributeDescriptorComparator.java]: remove duplicate header

File: catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/ColumnHeaderIterator.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+/*
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/ColumnHeaderIterator.java]: duplicate header

File: catalog/transformer/catalog-transformer-csv/src/test/java/ddf/catalog/transformer/csv/ColumnHeaderIteratorTest.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+/*
```
Comment (jlcsmith) [inline catalog/transformer/catalog-transformer-csv/src/test/java/ddf/catalog/transformer/csv/ColumnHeaderIteratorTest.java]: remove duplicate header

File: catalog/catalog-app/pom.xml (hunk 4 above)
Comment (dcruver) [inline catalog/catalog-app/pom.xml]: fixed

File: catalog/core/catalog-core-solrcommands/pom.xml (hunk 3 above)
Comment (dcruver) [inline catalog/core/catalog-core-solrcommands/pom.xml]: removed

File: catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/ColumnHeaderIterator.java (hunk 8 above)
Comment (dcruver) [inline catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/ColumnHeaderIterator.java]: removed

File: catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/CsvAttributeDescriptorComparator.java (hunk 7 above)
Comment (dcruver) [inline catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/CsvAttributeDescriptorComparator.java]: removed

File: catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformer.java (hunk 6 above)
Comment (dcruver) [inline catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformer.java]: removed

File: catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/MetacardIterator.java (hunk 5 above)
Comment (dcruver) [inline catalog/transformer/catalog-transformer-csv/src/main/java/ddf/catalog/transformer/csv/MetacardIterator.java]: removed

File: catalog/transformer/catalog-transformer-csv/src/test/java/ddf/catalog/transformer/csv/ColumnHeaderIteratorTest.java (hunk 9 above)
Comment (dcruver) [inline catalog/transformer/catalog-transformer-csv/src/test/java/ddf/catalog/transformer/csv/ColumnHeaderIteratorTest.java]: removed

**Comment by jlcsmith** [comment]
build now

**Comment by rzwiefel** [comment]
@dcruver Hey don, threw together some documentation for this real quick so we can push this through. 
Should just have to create a file under `distribution/docs/src/main/resources/_contents/_transformers/csv-query-response-transformer-contents.adoc` and then update `distribution/docs/src/main/resources/documentation.adoc` around line 540 (maybe underneath the 'xml-query-response-transformer-contents.adoc') to add an include for the previously created file.
https://gist.github.com/rzwiefel/d307a764f35e42657cbf38027753210f

**Comment by jlcsmith** [comment]
build now

File: catalog/transformer/catalog-transformer-csv/pom.xml [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Export-Package/>
+                        <Embed-Dependency>
+                            catalog-core-api-impl,
+                            commons-csv,
+                            commons-lang3,
```
Comment (kcwire) [inline catalog/transformer/catalog-transformer-csv/pom.xml]: commons-lang3 should be provided as a bundle - you shouldn't have to embed it.
https://github.com/codice/ddf/blob/master/platform/platform-app/src/main/resources/features.xml#L1081

**Comment by jlcsmith** [comment]
build now

File: pom.xml [hunk 11]
```diff
@@ -197,7 +197,6 @@
         <jodah-failsafe.version>0.9.5</jodah-failsafe.version>
         <apache.shiro.version>1.3.2</apache.shiro.version>
         <spark.version>2.5.5</spark.version>
-        <sardine.version>5.7</sardine.version>
```
Comment (jlcsmith) [inline pom.xml]: ❗️ @dcruver - why did this get removed?

**Comment by dcruver** [comment]
build now

**Comment by dcruver** [comment]
build now
